fusequery/query/src/datasources/local/memory_table.rs

```diff
@@ -0,0 +1,130 @@
+use common_infallible::RwLock;
+use common_planners::ReadDataSourcePlan;
+use common_planners::ScanPlan;
+use common_planners::Statistics;
+use common_planners::TableOptions;
+use common_streams::SendableDataBlockStream;
+use futures::stream::StreamExt;
+
+use super::MemoryTableStream;
+use crate::datasources::Common;
+use crate::datasources::Table;
+use crate::sessions::FuseQueryContextRef;
+
+pub struct MemoryTable {
+    db: String,
+    name: String,
+    schema: DataSchemaRef,
+    blocks: Arc<RwLock<Vec<DataBlock>>>,
+}
+
+impl MemoryTable {
+    pub fn try_create(
+        db: String,
+        name: String,
+        schema: DataSchemaRef,
+        _options: TableOptions,
+    ) -> Result<Box<dyn Table>> {
+        let table = Self {
+            db,
+            name,
+            schema,
+            blocks: Arc::new(RwLock::new(vec![])),
+        };
+        Ok(Box::new(table))
+    }
+}
+
+#[async_trait::async_trait]
+impl Table for MemoryTable {
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn engine(&self) -> &str {
+        "Memory"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> Result<DataSchemaRef> {
+        Ok(self.schema.clone())
+    }
+
+    fn is_local(&self) -> bool {
+        true
+    }
+
+    fn read_plan(
+        &self,
+        ctx: FuseQueryContextRef,
+        scan: &ScanPlan,
+        _partitions: usize,
+    ) -> Result<ReadDataSourcePlan> {
+        let blocks = self.blocks.read();
+        let rows = blocks.iter().map(|block| block.num_rows()).sum();
+        let bytes = blocks.iter().map(|block| block.memory_size()).sum();
+
+        Ok(ReadDataSourcePlan {
+            db: self.db.clone(),
+            table: self.name().to_string(),
+            schema: self.schema.clone(),
+            parts: Common::generate_parts(
+                0,
+                ctx.get_settings().get_max_threads()?,
+                blocks.len() as u64,
+            ),
+            statistics: Statistics::new_exact(rows, bytes),
+            description: format!("(Read from Memory Engine table  {}.{})", self.db, self.name),
+            scan_plan: Arc::new(scan.clone()),
+            remote: false,
+        })
+    }
+
+    async fn read(
+        &self,
+        ctx: FuseQueryContextRef,
+        _source_plan: &ReadDataSourcePlan,
+    ) -> Result<SendableDataBlockStream> {
+        let blocks = self.blocks.read();
+        Ok(Box::pin(MemoryTableStream::try_create(
+            ctx,
+            blocks.clone(),
+        )?))
+    }
+
+    async fn append_data(
+        &self,
+        _ctx: FuseQueryContextRef,
+        insert_plan: common_planners::InsertIntoPlan,
+    ) -> Result<()> {
+        let mut s = {
+            let mut inner = insert_plan.input_stream.lock().unwrap();
+            (*inner).take()
+        }
+        .ok_or_else(|| ErrorCode::EmptyData("input stream consumed"))?;
+
+        if insert_plan.schema().as_ref() != self.schema.as_ref() {
```

Comment:
Here we checked the schema,  seems it does not need to check in block stream.

Reply:
A little bit confused here, no sure if checking the equality of the table's schema and the insert_plan's schema is proper here. Does it mean we the caller of append_data should always provide full rows of memory_table?

Reply:
Oh, `memory_table` seems to be for test purposes only,  simplify the schema checking is ok for me.

And I suggest relaxing the schema checking in this PR and improve it in other PRs